src/libbson/src/bson/bson-context.c

```diff
@@ -264,79 +139,261 @@ _bson_context_get_hostname (char *out)
+
+   v1 ^= 0xdd;
 
-   /* Copy five random bytes, endianness does not matter. */
-   memcpy (&context->rand, (char *) &rand_bytes, sizeof (context->rand));
+   for (i = 0; i < D_ROUNDS; ++i)
+      _sip_round (&v0, &v1, &v2, &v3);
+
+   b = v0 ^ v1 ^ v2 ^ v3;
+   _u64_into_u8x8_le (digest_buf + 8, b);
+
+   memcpy (digest, digest_buf, sizeof digest_buf);
+}
+
+/*
+ * The seed consists of the following hashed together:
+ * - current time (with microsecond resolution)
+ * - current pid
+ * - current hostname
+ * - The init-call counter
+ */
+struct _init_rand_params {
+   struct timeval time;
+   uint64_t pid;
+   char hostname[HOST_NAME_MAX];
+   int64_t rand_call_counter;
+};
+
+static void
+_bson_context_init_random (bson_context_t *context, bool init_seq)
+{
+   /* Keep an atomic counter of this function being called. This is used to add
+    * additional input to the random hash, ensuring no two calls in a single
+    * process will receive identical hash inputs, even occurring at the same
+    * microsecond. */
+   static int64_t s_rand_call_counter = INT64_MIN;
+
+   /* The message digest of the random params */
+   uint64_t digest[2] = {0};
+   uint64_t key[2] = {0};
+   /* The randomness parameters */
+   struct _init_rand_params rand_params;
+
+   /* Init each part of the randomness source: */
+   memset (&rand_params, 0, sizeof rand_params);
+   bson_gettimeofday (&rand_params.time);
+   rand_params.pid = _bson_getpid ();
+   context->gethostname (rand_params.hostname);
+   rand_params.rand_call_counter = bson_atomic_int64_fetch_add (
+      &s_rand_call_counter, 1, bson_memory_order_seq_cst);
+
+   /* Generate a SipHash key. We do not care about secrecy or determinism, only
+    * uniqueness. */
+   memcpy (key, &rand_params, sizeof key);
+   key[1] = ~key[0];
+
+   /* Hash the param struct */
+   _siphash (&rand_params, sizeof rand_params, key, digest);
+
+   /** Initialize the rand and sequence counters with our random digest */
+   memcpy (context->randomness, digest, sizeof context->randomness);
+   if (init_seq) {
+      memcpy (&context->seq32, digest + 1, sizeof context->seq32);
+      memcpy (&context->seq64, digest + 1, sizeof context->seq64);
+      /* Chop off some initial bits for nicer counter behavior. This allows the
+       * low digit to start at a zero, and prevents immediately wrapping the
+       * counter in subsequent calls to set_oid_seq. */
+      context->seq32 &= ~UINT32_C (0xf0000f);
+      context->seq64 &= ~UINT64_C (0xf0000f);
+   }
+
+   /* Remember the PID we saw here. This may change in case of fork() */
+   context->pid = rand_params.pid;
 }
 
 static void
 _bson_context_init (bson_context_t *context, bson_context_flags_t flags)
 {
    context->flags = (int) flags;
-   context->oid_set_seq32 = _bson_context_set_oid_seq32;
-   context->oid_set_seq64 = _bson_context_set_oid_seq64;
    context->gethostname = _bson_context_get_hostname;
 
-   if ((flags & BSON_CONTEXT_THREAD_SAFE)) {
```

Comment:
I do not know why they were added originally. But are the thread unsafe variants of `_bson_context_set_oid_seq32` and `_bson_context_set_oid_seq64` removed because they provide no benefit?

Reply:
I believe so. The first introductory changeset they were still only differing in atomic operations versus regular operations. IMO, the cost of an atomic increment over a regular increment is so vanishingly small in the context of OID generation that it isn't worth having a non-threadsafe option. This is the main "context tweak" I mentioned. If we really want the non-threadsafe option to remain, I'm open to restoring it, but I think it's not worth the additional complexity. Removing the option will require docs updates and allows for further simplification by removing some function pointers in the context.

Thoughts?

Reply:
I do not know the motivation for the thread unsafe variants. If the atomic increment likely will not impact the default behavior, removing SGTM.